python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        LongType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [LongType, TimeType, TimestampType, TimestamptzType],
+        ),
+        StringType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == StringType,
+        ),
+        BinaryType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == BinaryType or isinstance(t, FixedType),
+        ),
+        UUIDType: (
+            lambda v: Bucket._MURMUR3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    @staticmethod
+    def _decimal_to_bytes(value: Decimal):
+        unscaled_value = Transform._unscale_decimal(value)
+        number_of_bytes = int(math.ceil(unscaled_value.bit_length() / 8))
+        return unscaled_value.to_bytes(length=number_of_bytes, byteorder="big")
+
+    def __init__(self, transform_type: Type, num_buckets: int):
+        if (
+            transform_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(transform_type, FixedType)
+            and not isinstance(transform_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {transform_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}",
+            f"transforms.bucket(transform_type={repr(transform_type)}, num_buckets={num_buckets})",
+        )
+        self._type = transform_type
+        self._num_buckets = num_buckets
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value):
+        if value is None:
+            return None
+
+        if isinstance(self._type, FixedType):
+            return (
+                Bucket._MURMUR3.hash(value) & Bucket._MAX_32_BITS_INT
+            ) % self._num_buckets
+        elif isinstance(self._type, DecimalType):
+            return (
+                Bucket._MURMUR3.hash(Bucket._decimal_to_bytes(value))
+                & Bucket._MAX_32_BITS_INT
+            ) % self._num_buckets
+
+        return (
+            Bucket._FUNCTIONS_MAP[self._type][0](value) & Bucket._MAX_32_BITS_INT
+        ) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        if isinstance(self._type, FixedType):
```

Comment:
I don't think that this logic should be copied from Java. The reason for it is that the same class (`BucketByteBuffer`) is used for both fixed and binary. Here the classes are different.

Reply:
Oh, I see. So from spec perspective, those two types are not transformable to each other?

Reply:
@rdblue so we won't mark bytes to bytearray as `can_transform` and verse vice?

Reply:
This is used so that we can sanity check whether concrete types are supported by a given Transform instance. It should be implemented for each instance of Transform, rather than copying the logic from other implementations.

Reply:
👌  updated it for python.